### PR TITLE
feat: add callback support to EstimatorTrial

### PR DESCRIPTION
**Pull request by aaron276h**
## Description
Most callback functionality can be handled via native support in tf.estimator using `tf.estimator.SessionRunHook`. However our checkpoint mechanism does use `CheckpointSaverHook`, thus checkpoint callbacks can not be performed via `tf.estimator.SessionRunHook`. To support such use cases (e.g., adding custom meta-data file to a checkpoint) we are adding support for deteremined callbacks in EstimatorTrial. Currently we are adding support for `on_checkpoint_load` and `on_checkpoint_end`.

Additionally as part of this PR, removed copies of checkpoint files that are no longer necessary. 

[DET-3100](https://determinedai.atlassian.net/browse/DET-3100)

## Test Plan
Added a unit test to test callback support. Also performed manual test confirming that changes do not break checkpoint/restart for multi-gpu.

### CODE REVIEW

File: harness/determined/estimator/_callback.py [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+    .. warning::
+        If distributed or parallel training is enabled, only the chief GPU (rank = 0)
+        will execute the `on_checkpoint_end` callbacks.
+    """
+
+    def on_checkpoint_load(self, checkpoint_dir: pathlib.Path) -> None:
+        """
+        Run when checkpoint is loaded at startup. If not resuming from checkpoint
+        this is never called.
+        """
```
Comment (yoavz) [inline harness/determined/estimator/_callback.py:18]: "at startup" -> "when the task environment container starts up"

File: harness/determined/estimator/_callback.py [hunk 2]
```diff
@@ -0,0 +1,29 @@
+import pathlib
+
+
+class EstimatorCallback:
+    """
+    Abstract base class used to define a callback that should execute during
+    the lifetime of a EstimatorCallback.
```
Comment (yoavz) [inline harness/determined/estimator/_callback.py:7]: blocking: add a warning that this callback should be stateless.

(in contrast, `PyTorchCallback` can specify how to save and load callback state and can be stateful)

**Comment by aaron276h** [comment]
@yoavz yes I definitely agree combining the callback interface with Estimator hooks definitely makes it a simpler UX and improves the native experience. I updated the PR with these changes.

File: harness/determined/estimator/_callback.py (hunk 1 above)
Comment (aaron276h) [reply harness/determined/estimator/_callback.py]: Done

File: harness/determined/estimator/_callback.py (hunk 2 above)
Comment (aaron276h) [reply harness/determined/estimator/_callback.py]: Users can save load/save callback state via `on_checkpoint_load` and `on_checkpoint_end` calls.

File: docs/reference/api/estimator.txt [hunk 3]
```diff
@@ -47,6 +47,18 @@ API or Native API.
...
+~~~~~~~~~
+
+To execute arbitrary Python functionality during the lifecycle of a
+``EstimatorTrial``, ``determined.estimator.RunHook`` extends
+`tf.estimator.SessionRunHook <https://www.tensorflow.org/api_docs/python/tf/estimator/SessionRunHook/>`_.
+
+.. autoclass:: determined.estimator.RunHook
+    :members: on_checkpoint_load, on_checkpoint_end
+
+
```
Comment (yoavz) [inline docs/reference/api/estimator.txt:61]: nonblocking: could be nice to add a code example here making it explicit that these are passed using `build_train_spec` and/or `build_valid_spec`

Comment (aaron276h) [reply docs/reference/api/estimator.txt:61]: Yep agreed, I added in an example.

**Pull request merged**